keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,504 @@
+
+At the beginning, Memory Manager initializes a collection of `Memory Table` for each NUMA node and memory type, which composes `Memory Map`. Below shows how `Memory Table` and `Memory Map` are structured in golang style.
+
+```go
+type MemoryTable struct {
+  TotalMemSize uint64
+  PreReserved  uint64
+  Allocatable  uint64
+  Reserved     uint64
+  Free         uint64
+}
+
+type MemoryMap map[NUMA_NODE_INDEX]map[MEMORY_TYPES]MemoryTable
+```
+_Figure: The construction of Memory Table and Memory Map._
+
+- For a brief example, on a two-socket machine with one size of pre-allocated hugepages, Memory Manager will generate four memory tables (two of NUMA nodes * two of memory types including a conventional memory type).
+
+`Memory Table` is proposed to represent amount of pre-reserved and allocatable memory for a certain memory type on a certain NUMA node. The table consists of _Pre-reserved Zone_ and _Allocatable Zone_. The following figure shows the structure of `Memory Table`, briefly.
+
+```
++-------------------+------------+--------------+
+|                   |     Allocatable Zone      |
++ Pre-reserved Zone +------------+--------------+
+|                   |  Reserved  |     Free     |
++-------------------+------------+--------------+
+```
+_Figure: The construction of Memory Table._
+
+The size of the two zones (_Pre-reserved Zone_ and _Allocatable Zone_) are pre-configured in kubelet configuration by administrator and are immutable values in runtime.
+- _Pre-reserved Zone_ indicates the certain amount of memory which is pre-reserved for system such as kernel, OS deamon, and core components of kubernetes like kubelet. For more details how to configure this size, see [The configuration of Memory Manager](#the-configuration-of-memory-manager) section.
+
+- _Allocatable Zone_ indicates the total size of allocatable memory for containers in Guaranteed QoS Pod. The size of this zone is simply the rest of  _Pre-reserved Zone_ from the total size of the Memory Table.
+
+- _Reserved_ section of _Allocatable Zone_ indicates the total size of reserved memory for already-deployed containers in Guaranteed QoS Pods. The Guaranteed memory of certain container is taken from Free memory section.
+
+- _Free_ section of _Allocatable Zone_ indicates a free memory for new containers in a Guaranteed QoS Pod. When a new container is deployed, _Free_ memory is subtracted by the requested memory size of the container. And when a running container is terminated, the requested memory size of the container is returned to _Free_ memory.
+
+#### Computing NUMA affinity
+
+Memory Manager calculates a topology hint for a container in Guaranteed QoS Pod for conventional memory and hugepages of all sizes. The topology hint is well known as _NUMA affinity_ which represents a possible NUMA node that has enough capacity of required memory for all memory types.
```

Comment:
the same here. topology hint is not the same as NUMA affinity.

Reply:
They are currently one and the same, but I agree, we should not think of the Hint as just the affinity. The affinity is just one part of the overall hint.